### PR TITLE
ENT-5553: Added valgrind test using Ubuntu 20 container

### PR DESCRIPTION
**Pull request by olehermanse**

### CODE REVIEW

**Review by cf-bottom — CHANGES_REQUESTED** [review]
Please use a company e-mail instead of o***@gmail.com

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5559/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5559/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5559/

File: tests/valgrind-check/valgrind.sh [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    echo "Stopping cfengine daemons"
+    pkill -f cf-serverd  || echo "Did not kill cf-serverd"
+    pkill -f cf-execd    || echo "Did not kill cf-execd"
+    pkill -f cf-monitord || echo "Did not kill cf-monitord"
+}
+
+function no_errors {
+    set +e
+    grep -i "error" $1
+    grep -i "error" $1 && exit 1
```
Comment (vpodzime) [inline tests/valgrind-check/valgrind.sh:23]: The second `grep` should use `-q`, maybe?

File: tests/valgrind-check/run_checks.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+set -x
+
+function check_with_valgrind() {
+  bash tests/valgrind-check/valgrind.sh
+}
+
+cd "$(dirname $0)"/../../
+
+failure=0
+check_with_valgrind || { echo "FAIL: valgrind check failed"; whoami; failure=1; }
```
Comment (vpodzime) [inline tests/valgrind-check/run_checks.sh:12]: I'd prefer `echo "FAIL: valgrind check failed [$(whoami)]"` or something like that.

File: tests/valgrind-check/run.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+set -e
+trap "echo FAILURE" ERR
+
+set -x
+
+cd ../../../
+
+if which podman ; then
+  CLI="podman --cgroup-manager=cgroupfs"
+  [ "`whoami`" == "root" ] || { echo "Error: Must be root (sudo) with podman" ; exit 1 ; }
```
Comment (vpodzime) [inline tests/valgrind-check/run.sh:12]: Please use good old `=` for "equals", this is not C. :)

File: tests/valgrind-check/valgrind.sh [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+function check_valgrind_output {
+    set -e
+    if [ ! -f "$1" ]; then
+        echo "$1 does not exists!"
+        exit 1
+    fi
+    echo "Looking for problems in $1:"
+    grep -i "ERROR SUMMARY: 0 errors" "$1"
+    cat $1 | sed -e "/ 0 errors/d" -e "/and suppressed error/d" -e "/a memory error detector/d" > filtered.txt
```
Comment (vpodzime) [inline tests/valgrind-check/valgrind.sh:40]: You can use `sed ... < $1` instead of `cat|sed`.

File: tests/valgrind-check/valgrind.sh [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    no_errors filtered.txt
+    set +e
+    grep -i "at 0x" filtered.txt
+    grep -i "at 0x" filtered.txt && exit 1
+    grep -i "by 0x" filtered.txt
+    grep -i "by 0x" filtered.txt && exit 1
+    grep -i "Failed to connect" filtered.txt
+    grep -i "Failed to connect" filtered.txt && exit 1
```
Comment (vpodzime) [inline tests/valgrind-check/valgrind.sh:48]: Same here with the `-q` option for the even `grep`s

File: tests/valgrind-check/valgrind.sh [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    set -e
+}
+
+function check_masterfiles_and_inputs {
+    set -e
+    echo "Comparing promises.cf from inputs and masterfiles:"
+    diff /var/cfengine/inputs/promises.cf /var/cfengine/masterfiles/promises.cf
```
Comment (vpodzime) [inline tests/valgrind-check/valgrind.sh:55]: Please use `diff -u`.

File: tests/valgrind-check/valgrind.sh [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+set -e
+set -x
+
+# Assume we are in core directory
+
+./autogen.sh --enable-debug
```
Comment (vpodzime) [inline tests/valgrind-check/valgrind.sh:63]: Please add `-C`, it speeds up `./configure` in *libntech* significantly.

File: tests/valgrind-check/valgrind.sh [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+sleep 10
+
+print_ps
+
+echo "Running cf-net:"
+valgrind $VG_OPTS /var/cfengine/bin/cf-net GET /var/cfengine/masterfiles/promises.cf 2>&1 | tee get.txt
+check_valgrind_output get.txt
+
+echo "Checking promises.cf diff (from cf-net GET):"
+diff ./promises.cf /var/cfengine/masterfiles/promises.cf
```
Comment (vpodzime) [inline tests/valgrind-check/valgrind.sh:112]: `diff -u` here as well

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise.

File: tests/valgrind-check/run.sh (hunk 3 above)
Comment (olehermanse) [reply tests/valgrind-check/run.sh]: This is the first result when googling "String equals bash":

https://linuxhint.com/compare_strings_bash/#:~:text=You%20can%20check%20the%20equality,two%20strings%20also%20in%20bash.

File: tests/valgrind-check/run_checks.sh (hunk 2 above)
Comment (olehermanse) [reply tests/valgrind-check/run_checks.sh:12]: This is just a leftover from debugging, removed.

File: tests/valgrind-check/valgrind.sh (hunk 6 above)
Comment (olehermanse) [reply tests/valgrind-check/valgrind.sh:55]: Added.

File: tests/valgrind-check/valgrind.sh (hunk 8 above)
Comment (olehermanse) [reply tests/valgrind-check/valgrind.sh:112]: Added.

File: tests/valgrind-check/valgrind.sh (hunk 7 above)
Comment (olehermanse) [reply tests/valgrind-check/valgrind.sh]: Done.

File: tests/valgrind-check/valgrind.sh (hunk 4 above)
Comment (olehermanse) [reply tests/valgrind-check/valgrind.sh:40]: I dunno, I think mine is more readable for non-wizards 🤷

File: tests/valgrind-check/valgrind.sh (hunk 1 above)
Comment (olehermanse) [reply tests/valgrind-check/valgrind.sh:23]: Yep, thanks.

File: tests/valgrind-check/valgrind.sh (hunk 5 above)
Comment (olehermanse) [reply tests/valgrind-check/valgrind.sh:48]: Added.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5562/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5562/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5562/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5564/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5564/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5564/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/5565/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/5565/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-5565/

**Review by vpodzime — APPROVED** [review]
Looks good to me.

**Comment by olehermanse** [comment]
Cherry-pick: https://github.com/cfengine/core/pull/4295

**Pull request merged**